db-scheduler/src/main/java/com/github/kagkarlsson/scheduler/task/schedule/DailyParser.java

```diff
@@ -0,0 +1,87 @@
+/**
+ * Copyright (C) Gustav Karlsson
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.github.kagkarlsson.scheduler.task.schedule;
+
+import com.github.kagkarlsson.scheduler.task.schedule.Schedules.UnrecognizableSchedule;
+
+import java.time.DateTimeException;
+import java.time.LocalTime;
+import java.time.ZoneId;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.function.Function;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+final class DailyParser implements Parser {
+    private static final Pattern DAILY_PATTERN_WITH_TIMEZONE = Pattern.compile("^DAILY\\|((\\d{2}:\\d{2})(,\\d{2}:\\d{2})*)(\\|(.+))?$");
+    private static final List<String> EXAMPLES = Arrays.asList("DAILY|12:00", "DAILY|12:00,13:45", "DAILY|12:00,13:45|Europe/Rome");
+
+    @Override
+    public Schedule parse(String scheduleString) {
+        MatchedSchedule matchedSchedule = match(scheduleString);
+        return matchedSchedule.maybeZoneId()
+            .map(zoneId -> new Daily(zoneId, matchedSchedule.scheduleTimes()))
+            .orElse(new Daily(matchedSchedule.scheduleTimes()));
+    }
+
+    private MatchedSchedule match(String scheduleString) {
+        return Optional.ofNullable(scheduleString)
+            .map(DAILY_PATTERN_WITH_TIMEZONE::matcher)
+            .filter(Matcher::matches)
+            .map(MatchedSchedule::new)
+            .orElseThrow(() -> new UnrecognizableSchedule(scheduleString, this.examples()));
+    }
```

Comment:
Have you considered making an abstract parent class that might dry things up even more? Have a feeling all parsers will be regex-based..

```
public abstract class RegexParser implements Parser {
....
```

Reply:
I've tried extracting the shared regex related logic to a `OptionalMatcher` class, what do you think?

Reply:
I still feel that the code might be clearer if you structure the code similar to this:

```
public abstract class RegexBasedParser implements Parser {
    private final Pattern pattern;
    private final List<String> examples;

    public RegexBasedParser(Pattern pattern, List<String> examples) {
        this.pattern = pattern;
        this.examples = examples;
    }

    @Override
    public Optional<Schedule> parse(String scheduleString) {
        Matcher m = pattern.matcher(scheduleString);
        if (m.matches()) {
            return getMatchedSchedule(m);
        } else {
            return Optional.empty();
        }
    }

    @Override
    public List<String> examples() {
        return examples;
    }

    public abstract Optional<Schedule> getMatchedSchedule(MatchResult result);

}

...
(example re-implementation of FixedDelayParser:)

final class FixedDelay2Parser extends RegexBasedParser {
    private static final Pattern FIXED_DELAY_PATTERN = Pattern.compile("^FIXED_DELAY\\|(\\d+)s$");
    private static final List<String> EXAMPLES = Collections.singletonList("FIXED_DELAY|120s");

    FixedDelay2Parser() {
        super(FIXED_DELAY_PATTERN, EXAMPLES);
    }

    @Override
    public Optional<Schedule> getMatchedSchedule(MatchResult result) {
        FixedDelay schedule = FixedDelay.ofSeconds(Integer.parseInt(result.group(1)));
        return Optional.of(schedule);
    }
}
```

Reply:
Should be clearer now :relaxed:

Reply:
It should look clearer now :relieved: